datasource/src/main/scala/quasar/physical/mongo/interpreter/Mask.scala

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo.interpreter
+
+import slamdata.Predef._
+
+import cats.syntax.order._
+import cats.syntax.foldable._
+import cats.instances.option._
+import cats.instances.list._
+
+import quasar.common.CPath
+import quasar.api.table.ColumnType
+
+import quasar.physical.mongo.MongoExpression
+import quasar.physical.mongo.{Aggregator, Version, MongoExpression => E}
+
+import shims._
+
+object Mask {
+  final case class TypeTree(types: Set[ColumnType], obj: Map[String, TypeTree], list: Map[Int, TypeTree])
```

Comment:
This is for filtering, e.g. we have `foo.bar.baz: Int, foo.quux: String, foo[0]zoo: Object`. Instead of matching/filtering this creates an object expression that is set as a value of `$project` field and value of `foo` is set as either `null|undefined` or if (typeExpr foo) is object or array object or array with recursion results.